src/mlpack/methods/ann/layer/virtual_batch_norm_impl.hpp

```diff
@@ -0,0 +1,148 @@
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_VIRTUALBATCHNORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_VIRTUALBATCHNORM_IMPL_HPP
+
+// In case it is not included.
+#include "virtual_batch_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+VirtualBatchNorm<InputDataType, OutputDataType>::VirtualBatchNorm() :
+    size(0),
+    eps(1e-8),
+    loading(false),
+    oldCoefficient(0),
+    newCoefficient(0)
+{
+  // Nothing to do here.
+}
+template <typename InputDataType, typename OutputDataType>
+template<typename eT>
+VirtualBatchNorm<InputDataType, OutputDataType>::VirtualBatchNorm(
+    const arma::Mat<eT>& referenceBatch,
+    const size_t size,
+    const double eps) :
+    size(size),
+    eps(eps),
+    loading(false)
+{
+  weights.set_size(size + size, 1);
+
+  referenceBatchMean = arma::mean(referenceBatch, 1);
+  referenceBatchMeanSquared = arma::mean(arma::square(referenceBatch), 1);
+  newCoefficient = 1.0 / (referenceBatch.n_cols + 1);
+  oldCoefficient = 1 - newCoefficient;
+}
+
+template<typename InputDataType, typename OutputDataType>
+void VirtualBatchNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size, 1, false, false);
+
+  if (!loading)
+  {
+    gamma.fill(1.0);
+    beta.fill(0.0);
+  }
+  loading = false;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void VirtualBatchNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  inputParameter = input;
+  arma::mat inputMean = arma::mean(input, 1);
+  arma::mat inputMeanSquared = arma::mean(arma::square(input), 1);
+
+  mean = oldCoefficient * referenceBatchMean + newCoefficient * inputMean;
+  arma::mat meanSquared = oldCoefficient * referenceBatchMeanSquared +
+      newCoefficient * inputMeanSquared;
+  variance = meanSquared - arma::square(mean);
+  // Normalize the input.
+  output = input.each_col() - mean;
+  temp = output;
+  output.each_col() /= arma::sqrt(variance + eps);
+
+  // Reused in the backward and gradient step.
+  normalized = output;
+  // Scale and shift the output.
+  output.each_col() %= gamma;
+  output.each_col() += beta;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void VirtualBatchNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  // Step 1: dl / dxhat.
```

Comment:
Could you please ad this punctuation to `BatchNorm` and `LayerNorm` code as well?